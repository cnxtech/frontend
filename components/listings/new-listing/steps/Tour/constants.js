export const NEXT = 1
export const PREVIOUS = -1

export const MAX_MONTHS_TO_DISPLAY = 1
export const MAX_DAYS_TO_DISPLAY = 3

export const EARLY = '09'
export const LATE = '17'
export const EARLY_DISPLAY = 'No come√ßo do dia (9:00h)'
export const LATE_DISPLAY = 'No final do dia (17:00h)'
