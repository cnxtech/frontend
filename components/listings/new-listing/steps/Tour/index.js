import React, { Fragment, Component } from 'react'
import { Formik, Field } from 'formik'

import RadioButton from '@emcasa/ui-dom/components/RadioButton'
import Row from '@emcasa/ui-dom/components/Row'
import Col from '@emcasa/ui-dom/components/Col'
import View from '@emcasa/ui-dom/components/View'
import NavButtons from 'components/listings/new-listing/shared/NavButtons'
import CustomTime from './components/CustomTime'
import TourMonths from './components/TourMonths'
import TourDays from './components/TourDays'
import {
  getTourDays,
  getTourMonths,
  getTimeDisplay,
  TOUR_HOURS,
  EARLY,
  LATE,
} from 'components/listings/new-listing/lib/times'

class Tour extends Component {
  constructor(props) {
    super(props)
    this.done = this.done.bind(this)
    this.previousStep = this.previousStep.bind(this)

    this.onMonthChanged = this.onMonthChanged.bind(this)
    this.onDaySelected = this.onDaySelected.bind(this)

    this.selectTime = this.selectTime.bind(this)
    this.hasCustomTime = this.hasCustomTime.bind(this)
    this.selectCustomTime = this.selectCustomTime.bind(this)
  }

  state = {
    month: null,
    day: null,
    time: null,
    customTime: false,
    monthOffset: 0,
    dayOffset: 0
  }

  componentDidMount() {
    this.updateStateFromProps(this.props)
  }

  componentWillReceiveProps(props) {
    this.updateStateFromProps(props)
  }

  updateStateFromProps(props) {
    const { tour, services } = props
    let firstAvailableMonth
    if (services) {
      firstAvailableMonth = getTourMonths(services.tourOptions)[0]
    }
    if (tour) {
      this.setState({
        month: tour.month || firstAvailableMonth.date.getMonth(),
        day: tour.day,
        time: tour.time,
        customTime: tour.customTime,
        dayOffset: tour.dayOffset,
        monthOffset: tour.monthOffset
      })
    }
  }

  done() {
    const { navigateTo, updateTour } = this.props
    updateTour(this.state)
    navigateTo('services')
  }

  previousStep() {
    this.props.navigateTo('services')
  }

  onMonthChanged(month, monthOffset) {
    this.setState({
      month,
      monthOffset
    })
  }

  onDaySelected(day, dayOffset, setFieldTouched, setFieldValue) {
    setFieldValue('day', day)
    setFieldTouched('day')
    setFieldValue('time', null)
    setFieldTouched('time')
    this.setState({
      day,
      dayOffset,
      time: null
    })
  }

  selectTime(value) {
    this.setState({time: value})
  }

  validateDay(value) {
    if (!value) {
      return 'É necessário selecionar um dia.'
    }
  }

  validateTime(value) {
    if (!value) {
      return 'É necessário selecionar um horário.'
    }
  }

  hasCustomTime(tourHours) {
    if (tourHours.length === 0) {
      return false
    }
    return !(tourHours.includes(EARLY) && tourHours.includes(LATE) && tourHours.length === 2)
  }

  selectCustomTime() {
    const { time, customTime } = this.state
    if (time) {
      this.setState({
        time: null,
        customTime: true
      })
      return
    }

    if (!customTime) {
      this.setState({
        time: null
      })
    }
    this.setState({
      customTime: true
    })
  }

  render() {
    const { tour, services } = this.props
    let month, day, time, monthOffset, dayOffset
    if (tour) {
      month = tour.month
      day = tour.day
      time = tour.time
      monthOffset = tour.monthOffset
      dayOffset = tour.dayOffset
    }
    const { tourOptions } = services
    const tourDays = getTourDays(tourOptions, this.state.month)
    const tourMonths = getTourMonths(tourOptions)

    return (
      <div ref={this.props.hostRef}>
        <Row justifyContent="center" p={4}>
          <Col width={[1, 1/2]}>
            <Formik
              initialValues={{
                month,
                day,
                time
              }}
              isInitialValid={() => {
                return !this.validateTime(time) || !this.validateDay(day)
              }}
              render={({isValid, setFieldTouched, setFieldValue}) => (
                <>
                  <Row mb={4}>
                    <Field
                      name="month"
                      render={() =>
                        <TourMonths
                          initialMonthOffset={monthOffset}
                          tourMonths={tourMonths}
                          onMonthChanged={this.onMonthChanged}
                        />
                      }/>
                  </Row>
                  <Row mb={4}>
                    <Field
                      name="day"
                      validate={this.validateDay}
                      render={() =>
                        <TourDays
                          month={month}
                          initialDayOffset={dayOffset || this.state.dayOffset}
                          initialDay={day}
                          tourDays={tourDays}
                          onDaySelected={(day, dayOffset) => {
                            this.onDaySelected(day, dayOffset, setFieldTouched, setFieldValue)
                          }}
                        />
                      }/>
                  </Row>
                  <Row mb={4}>
                    <Field
                      name="time"
                      validate={this.validateTime}
                      render={() =>
                        <Col width={1}>
                          <RadioButton.Group>
                            {TOUR_HOURS.map((item, index) => {
                              if (item !== EARLY && item !== LATE) {
                                return null
                              }
                              return (
                                <Fragment key={index}>
                                  <RadioButton
                                    label={getTimeDisplay(item, true)}
                                    value={item}
                                    checked={this.state.time === item && !this.state.customTime}
                                    onClick={() => {
                                      setFieldValue('time', item)
                                      setFieldTouched('time')
                                      this.setState({customTime: false})
                                      this.selectTime(item)
                                    }}
                                  />
                                  <View mb={2} />
                                </Fragment>
                              )
                            })}
                          </RadioButton.Group>
                          {this.hasCustomTime(TOUR_HOURS) &&
                            <CustomTime
                              onClick={this.selectCustomTime}
                              selected={this.state.customTime}
                              selectedTime={this.state.time}
                            >
                              {TOUR_HOURS.map((item, index) => {
                                return (
                                  <Fragment key={index}>
                                    <CustomTime.Item
                                      onClick={() => {
                                        setFieldValue('time', item)
                                        setFieldTouched('time')
                                        this.selectTime(item)
                                      }}
                                    >{getTimeDisplay(item, false)}</CustomTime.Item>
                                    <View mb={2} mr={2} />
                                  </Fragment>
                                )
                              })}
                            </CustomTime>}
                        </Col>
                      }/>
                  </Row>
                  <NavButtons
                    previousStep={this.previousStep}
                    onSubmit={this.done}
                    submitEnabled={isValid}
                  />
                </>
              )}
            />
          </Col>
        </Row>
      </div>
    )
  }
}

export default Tour
