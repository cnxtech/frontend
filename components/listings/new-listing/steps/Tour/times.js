import moment from 'moment'

const MONTH_KEY_FORMAT = 'YYYY-MM'
const DAY_KEY_FORMAT = 'YYYY-MM-DD'

const getTourMonths = (timeList) => {
  const tourMonths = {}
  timeList.forEach((item) => {
    const parser = moment(item)
    const display = parser.format('MMMM [de] YYYY')
    const key = parser.format(MONTH_KEY_FORMAT)

    if (!tourMonths[key]) {
      tourMonths[key] = {
        date: new Date(item),
        display: display
      }
    }
  })
  return tourMonths
}

const getTourDays = (timeList, month) => {
  const tourDays = {}
  timeList.forEach((item) => {
    const date = new Date(item)
    if (date.getMonth() !== month) return

    const parser = moment(item)
    const key = parser.format(DAY_KEY_FORMAT)

    if (!tourDays[key]) {
      tourDays[key] = {
        date,
        day: parser.format('DD'),
        dayOfWeek: parser.format('ddd')
      }
    }
  })
  return tourDays
}

const getTourHours = (timeList, date) => {
  const tourHours = []
  timeList.forEach((item) => {
    const parser = moment(item)
    const itemDate = parser.format(DAY_KEY_FORMAT)
    if (date !== itemDate) return
    tourHours.push(parser.format('HH'))
  })
  return tourHours
}

export {
  getTourMonths,
  getTourDays,
  getTourHours
}