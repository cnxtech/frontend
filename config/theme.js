import theme from '@emcasa/ui'
import {fas} from '@fortawesome/pro-solid-svg-icons'
import {fal} from '@fortawesome/pro-light-svg-icons'
import icons from './icons'

export default {
  ...theme,
  icons: {emcasa: icons, default: fas, light: fal}
}
