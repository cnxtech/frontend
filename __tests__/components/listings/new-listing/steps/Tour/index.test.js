import React from 'react';
import Tour from 'components/listings/new-listing/steps/Tour';
import renderer from 'react-test-renderer';
import 'jest-styled-components'
import theme from '@emcasa/ui'
import { ThemeProvider } from 'styled-components'

it('renders correctly', () => {
  const tree = renderer.create(
    <ThemeProvider theme={theme}>
      <Tour
        services={{
          tourOptions: [
            '2018-11-16T09:15:57.739175',
            '2018-11-16T17:15:57.739175',
            '2018-11-15T09:15:57.739175',
            '2018-11-15T17:15:57.739175',
            '2018-11-14T09:15:57.739175',
            '2018-11-14T17:15:57.739175',
            '2018-11-13T09:15:57.739175',
            '2018-11-13T17:15:57.739175',
            '2018-11-12T09:15:57.739175',
            '2018-11-12T17:15:57.739175'
          ]
      }}
      />
    </ThemeProvider>).toJSON()
  expect(tree).toMatchSnapshot()
})
