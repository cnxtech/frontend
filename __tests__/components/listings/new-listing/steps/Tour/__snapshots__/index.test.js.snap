// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c14 {
  margin-bottom: 10px;
}

.c18 {
  margin-right: 14px;
}

.c11 {
  box-sizing: border-box;
  cursor: pointer;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  padding: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 60px;
}

.c12 {
  box-sizing: border-box;
  width: 18px;
  height: 18px;
  border-radius: 18px;
  border: 2px solid #979797;
  background-color: transparent;
  margin-right: 14px;
}

.c13 {
  -webkit-font-smoothing: antialiased;
  cursor: pointer;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-family: FaktSoftPro-Normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 20px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c22 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  width: 100%;
}

.c4 {
  margin-right: 20px;
}

.c9 {
  margin-left: 20px;
}

.c10 {
  width: 100%;
}

.c23 {
  width: 41.66666666666667%;
}

.c6 {
  outline: none;
  cursor: default;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  background-color: #ffffff;
  width: auto;
  height: 40px;
  color: #8492a6;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c6 :hover {
  background-color: #ffffff;
}

.c24 {
  outline: none;
  display: block;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  background-color: #ffffff;
  width: 100%;
  height: 60px;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c24 :hover {
  background-color: #ffffff;
}

.c25 {
  outline: none;
  display: block;
  cursor: default;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  background-color: #ffffff;
  width: 100%;
  height: 60px;
  color: #8492a6;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c25 :hover {
  background-color: #ffffff;
}

.c8 {
  -webkit-font-smoothing: antialiased;
  font-family: FaktSoftPro-Normal;
  color: #3c4858;
  font-size: 18px;
}

.c20 {
  position: absolute;
  box-sizing: border-box;
  margin-top: 10px;
  margin-left: 10px;
  border-radius: 50%;
  border: 2px solid #979797;
  width: 20px;
  height: 20px;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid #e0e6ed;
  border-radius: 4px;
  cursor: pointer;
  padding-left: 10px;
  padding-right: 10px;
  min-height: 60px;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 10px;
  margin-bottom: 10px;
}

.c19 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  width: 18px;
  height: 18px;
}

.c21 {
  font-size: 16px;
}

.c5 {
  border: none;
  padding: 0;
  margin: 0;
}

@media screen and (min-width:640px) {
  .c1 {
    width: 50%;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
      width={
        Array [
          1,
          0.5,
        ]
      }
    >
      <div
        className="c2"
      >
        <div
          className="c3"
          width={1}
        >
          <div
            className="c4"
          >
            <button
              className="c5 c6"
              disabled={true}
              fontSize="default"
              height="medium"
              onClick={[Function]}
            >
              <div
                className=""
              >
                <svg
                  height={20}
                  version="1.1"
                  viewBox="0 0 192 512"
                  x={0}
                  y={0}
                >
                  <path
                    d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
                    fill="#979797"
                    name="caret-left"
                    type="default"
                  />
                </svg>
              </div>
            </button>
          </div>
          <div
            className=""
          >
            <div
              className="c7"
            >
              <div
                className=""
              >
                <p
                  className="c8"
                  color="dark"
                  data={
                    Object {
                      "date": 2018-11-16T11:15:57.739Z,
                      "display": "Novembro de 2018",
                      "key": "2018-11",
                    }
                  }
                  fontSize="default"
                  noBorder={true}
                >
                  Novembro de 2018
                </p>
              </div>
            </div>
          </div>
          <div
            className="c9"
          >
            <button
              className="c5 c6"
              disabled={true}
              fontSize="default"
              height="medium"
              onClick={[Function]}
            >
              <div
                className=""
              >
                <svg
                  height={20}
                  version="1.1"
                  viewBox="0 0 192 512"
                  x={0}
                  y={0}
                >
                  <path
                    d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
                    fill="#979797"
                    name="caret-right"
                    type="default"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        className="c2"
      >
        <div
          className="c3"
          width={1}
        >
          <div
            className="c4"
          >
            <button
              className="c5 c6"
              disabled={true}
              fontSize="default"
              height="medium"
              onClick={[Function]}
            >
              <div
                className=""
              >
                <svg
                  height={20}
                  version="1.1"
                  viewBox="0 0 192 512"
                  x={0}
                  y={0}
                >
                  <path
                    d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
                    fill="#979797"
                    name="caret-left"
                    type="default"
                  />
                </svg>
              </div>
            </button>
          </div>
          <div
            className=""
          >
            <div
              className="c7"
            >
              <div
                className=""
              />
            </div>
          </div>
          <div
            className="c9"
          >
            <button
              className="c5 c6"
              disabled={true}
              fontSize="default"
              height="medium"
              onClick={[Function]}
            >
              <div
                className=""
              >
                <svg
                  height={20}
                  version="1.1"
                  viewBox="0 0 192 512"
                  x={0}
                  y={0}
                >
                  <path
                    d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
                    fill="#979797"
                    name="caret-right"
                    type="default"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        className="c2"
      >
        <div
          className="c10"
          width={1}
        >
          <div
            className=""
            onChange={[Function]}
          >
            <div
              checked={false}
              className="c11"
              fontSize="default"
              height="tall"
              letterSpacing="default"
              onClick={[Function]}
              value="09"
            >
              <div
                checked={false}
                className="c12"
                label="No começo do dia (das 09h às 12h)"
              />
              <label
                className="c13"
                onClick={[Function]}
                value="09"
              >
                No começo do dia (das 09h às 12h)
              </label>
            </div>
            <div
              className="c14"
            />
            <div
              checked={false}
              className="c11"
              fontSize="default"
              height="tall"
              letterSpacing="default"
              onClick={[Function]}
              value="17"
            >
              <div
                checked={false}
                className="c12"
                label="No final do dia (das 16h às 19h)"
              />
              <label
                className="c13"
                onClick={[Function]}
                value="17"
              >
                No final do dia (das 16h às 19h)
              </label>
            </div>
            <div
              className="c14"
            />
          </div>
          <div
            className="c15 c16"
            onClick={[Function]}
            selected={false}
          >
            <div
              className="c10"
              width={1}
            >
              <div
                className="c16"
              >
                <div
                  className="c10"
                  width={1}
                >
                  <div
                    className="c17 c16"
                  >
                    <div
                      className="c18"
                    >
                      <div
                        className="c19 c20"
                      />
                    </div>
                    <span
                      className="c21 c8"
                      color="dark"
                      fontSize="default"
                    >
                      Escolher horário específico
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="c22"
      >
        <div
          className="c23"
          width={0.4166666666666667}
        >
          <button
            className="c24"
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Voltar
          </button>
        </div>
        <div
          className="c23"
          width={0.4166666666666667}
        >
          <button
            className="c25"
            disabled={true}
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Avançar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
